_posts/2024-04-20-coroutines.md

```diff
@@ -0,0 +1,348 @@
+ 2) use threads for the processes
+
+Let's take communication over UART as an example:
+ 1) iterative process means the usage of structure to store the state of the communication, and provide a function that based on the current state advances the state. (Commonly with the usage of an enum representing actual status)
+ 2) implement the communication in a thread which might release its time if it waits for some operation
+
+Approach 1) has multiple potential issues, it might take a lot of code to implement a complex exchange of data in this way (a lot of state variables) and the approach is problematic, as there is a chance that one of the steps might take longer than expected and we can't prevent that.
+
+Approach 2) has another set of issues. Each thread requires its own stack space (which might not scale), and we got all the problems of parallelism - exchanging data between can suffer various potential concurrency issues.
+
+From this perspective, coroutines are a third way of approaching these processes, which is not better than 1) or 2), but also not worse.
+Coroutines bring in a new (and not yet finetuned) mechanism that makes it possible to write an exchange of data over UART in a way that does not require as complex code as 1), does not suffer so many concurrency issues as 2) (and does not need its own stack, just frame).
+
+However, coroutines still suffer from the issue that one step of computation might take longer than expected, and we can assume that any errors in coroutines might be harder to inspect.
+
+What is interesting, is that coroutines have good potential to have more efficient context switches than threads, which was shown in this article: https://www.researchgate.net/publication/339221041_C20_Coroutines_on_Microcontrollers_-What_We_Learned
+
+Note that the article does not compare coroutines in a variety of situations, but it still shows something.
+
+### Problems
+
+There are multiple potential issues with coroutines and using them correctly, but I think there are a few that are noteworthy for the embedded community.
+
+#### Dynamic memory
+
+The first thing that I assume most of you noticed at the beginning is the fact the coroutine requires dynamic memory.
+This is not favourable in embedded in many cases, but there are ways to avoid that.
+
+Coroutines have `allocator` support, we can provide the coroutine with an allocator that can be used to get memory for the coroutine and hence avoid the dynamic allocation. (Approach that I can suggest)
+This is done by implementing custom `operator new` and `operator delete` on the `promise_type` which allocates the `entire frame`.
+
+An alternative is to rely on `halo` optimization (Heap Allocation Elision Optimization) if the coroutine is implemented correctly and the parent function executes the entire coroutine in its context.
+The compiler can optimize away the dynamic allocation and just store the frame on the stack of the coroutine's parent.
```

Comment:
To me this seems cleaner than a custom allocator. Could you elaborate on why you prefer the custom allocator solution?